src/core/crypto/relay_crypto.c

```diff
@@ -142,6 +142,13 @@ relay_decrypt_cell(circuit_t *circ, cell_t *cell,
           if (relay_digest_matches(thishop->crypto.b_digest, cell)) {
             *recognized = 1;
             *layer_hint = thishop;
+            /* Keep current digest of this cell for the possible SENDME. */
+            if (thishop->crypto.sendme_digest) {
+              crypto_digest_free(thishop->crypto.sendme_digest);
+            }
+            thishop->crypto.sendme_digest =
+              crypto_digest_dup(thishop->crypto.b_digest);
```

Comment:
This approach seems needlessly wasteful to me.  The digest object is as large as the entire internal digest object's state, which is often much larger than the actual set of bytes you're transmitting.  Why not just store the digest itself, rather than the crypto_digest_t?

Reply:
Done in commit 3943d096cc060d7676e1549e37d9577951b3b68f